### PR TITLE
[PyTorch] Add Anchor Box (New)

### PR DESCRIPTION
**Pull request by suvaansh**
*Description of changes:*

By submitting this pull request, I confirm that you can use, modify,
copy, and redistribute this contribution, under the terms of your
choice.

### CODE REVIEW

**Comment by mli** [comment]
Job d2l-en/PR-1370/4 is complete. 
Check the results at http://preview.d2l.ai/d2l-en/PR-1370/

File: chapter_computer-vision/anchor.md [hunk 1]
```diff
@@ -210,9 +413,77 @@ output = npx.multibox_detection(
...
+    return filtered_bb
+```
+
+```{.python .input}
+#@tab pytorch
+#@save
+def multibox_detection(id_cat, cls_probs, anchors, nms_threshold):
+
+    from collections import namedtuple
+    PredBoundingBox = namedtuple("PredBoundingBox", ["probability", "class_id", "classname", "bounding_box"])
```
[the remark this answers is not in the document]
Comment (AnirudhDagar) [reply chapter_computer-vision/anchor.md:443]: @suvaansh why don't we use `torchvision.ops.nms` here?
https://pytorch.org/docs/stable/torchvision/ops.html#torchvision.ops.nms

**Comment by AnirudhDagar** [comment]
@astonzhang @mli Could you please take a look at this PR before we merge it? I've requested a few changes from @suvaansh  Overall LGTM!

File: chapter_computer-vision/anchor.md [hunk 2]
```diff
@@ -36,16 +45,54 @@ Y = npx.multibox_prior(X, sizes=[0.75, 0.5, 0.25], ratios=[1, 2, 0.5])
...
+    scale = feature_map_sizes
+    steps_y = [1 / scale[0]]
+    steps_x = [1 / scale[1]]
+    sizes = [s / max(scale) for s in sizes]
+    num_layers = 1
+    boxes = []
+    for i in range(num_layers):
+        for h, w in itertools.product(range(feature_map_sizes[0]), range(feature_map_sizes[1])):
+            cx = (w + 0.5)*steps_x[i]
+            cy = (h + 0.5)*steps_y[i]
```
Comment (astonzhang) [inline chapter_computer-vision/anchor.md:63]: put space around operators *

File: chapter_computer-vision/anchor.md [hunk 3]
```diff
@@ -36,16 +45,54 @@ Y = npx.multibox_prior(X, sizes=[0.75, 0.5, 0.25], ratios=[1, 2, 0.5])
...
+                boxes.append((cx, cy, (s * math.sqrt(ar)), (s / math.sqrt(ar))))
+    return d2l.tensor(boxes)
+```
+
+```{.python .input}
+#@tab pytorch
+img = d2l.plt.imread('../img/catdog.jpg')
+h, w = img.shape[0:2]
+
+print(h,w)
```
Comment (astonzhang) [inline chapter_computer-vision/anchor.md:79]: put a space after comma

File: chapter_computer-vision/anchor.md [hunk 4]
```diff
@@ -79,6 +151,32 @@ show_bboxes(fig.axes, boxes[250, 250, :, :] * bbox_scale,
              's=0.75, r=0.5'])
 ```
 
+```{.python .input}
+#@tab pytorch
+#@save
+def center_2_hw(box: torch.Tensor) -> float:
```
Comment (astonzhang) [inline chapter_computer-vision/anchor.md:157]: why do we save center_2_hw to d2l?

File: chapter_computer-vision/anchor.md [hunk 5]
```diff
@@ -91,6 +189,48 @@ In fact, we can consider the pixel area of a bounding box as a collection of pix
 ![IoU is the ratio of the intersecting area to the union area of two bounding boxes.  ](../img/iou.svg)
 :label:`fig_iou`
 
+```{.python .input}
+#@tab pytorch
+#@save
```
Comment (astonzhang) [inline chapter_computer-vision/anchor.md:194]: why do we save intersect to d2l?

File: chapter_computer-vision/anchor.md [hunk 6]
```diff
@@ -36,16 +45,54 @@ Y = npx.multibox_prior(X, sizes=[0.75, 0.5, 0.25], ratios=[1, 2, 0.5])
 Y.shape
 ```
 
+```{.python .input}
+#@tab pytorch
+#@save
+def multibox_prior(feature_map_sizes, sizes, aspect_ratios):
```
Comment (astonzhang) [inline chapter_computer-vision/anchor.md:51]: why do we save multibox_prior to d2l?

File: chapter_computer-vision/anchor.md [hunk 7]
```diff
@@ -68,9 +115,34 @@ def show_bboxes(axes, bboxes, labels=None, colors=None):
                       bbox=dict(facecolor=color, lw=0))
 ```
 
+```{.python .input}
+#@tab pytorch
+#@save
+
```
Comment (astonzhang) [inline chapter_computer-vision/anchor.md:121]: remove this empty line

File: chapter_computer-vision/anchor.md [hunk 8]
```diff
@@ -91,6 +189,48 @@ In fact, we can consider the pixel area of a bounding box as a collection of pix
 ![IoU is the ratio of the intersecting area to the union area of two bounding boxes.  ](../img/iou.svg)
 :label:`fig_iou`
 
+```{.python .input}
+#@tab pytorch
+#@save
+def intersect(box_a: torch.Tensor, box_b: torch.Tensor) -> float:
+    # Coverting (cx, cy, w, h) to (x1, y1, x2, y2) since its easier to extract min/max coordinates
```
[the remark this answers is not in the document]
Comment (astonzhang) [reply chapter_computer-vision/anchor.md:196]: should be 78

File: chapter_computer-vision/anchor.md [hunk 9]
```diff
@@ -91,6 +189,48 @@ In fact, we can consider the pixel area of a bounding box as a collection of pix
...
+    return box[:, 2] * box[:, 3]
+
+def jaccard(box_a: torch.Tensor, box_b: torch.Tensor) -> float:
+#     print(box_a.shape)
+    intersection = intersect(box_a, box_b)
+    union = box_area(box_a).unsqueeze(1) + box_area(box_b).unsqueeze(0) - intersection
+    return intersection / union
+
+def find_overlap(bb_true_i, anchors, jaccard_overlap):
+
```
Comment (astonzhang) [inline chapter_computer-vision/anchor.md:217]: plz remove unnecessary empty lines within a function (unless it improves readability)

File: chapter_computer-vision/anchor.md [hunk 10]
```diff
@@ -140,20 +280,64 @@ anchors = np.array([[0, 0.1, 0.2, 0.3], [0.15, 0.2, 0.4, 0.4],
...
+    for j in range(len(overlap_list)):
+        overlap = overlap_list[j]
+        class_target[overlap] = class_true[j, 0].long()
+        overlap_coordinates[overlap] = 1.
+
+    new_anchors = torch.cat([*anchors])
+    overlap_coordinates = torch.cat([*overlap_coordinates])
+    new_anchors = new_anchors*overlap_coordinates
+
+    return (new_anchors.unsqueeze(0), overlap_coordinates.unsqueeze(0), class_target.unsqueeze(0))
```
[the remark this answers is not in the document]
Comment (astonzhang) [reply chapter_computer-vision/anchor.md:329]: should be 78

File: chapter_computer-vision/anchor.md [hunk 11]
```diff
@@ -140,20 +280,64 @@ anchors = np.array([[0, 0.1, 0.2, 0.3], [0.15, 0.2, 0.4, 0.4],
...
-```{.python .input  n=7}
+```{.python .input}
 labels = npx.multibox_target(np.expand_dims(anchors, axis=0),
                              np.expand_dims(ground_truth, axis=0),
                              np.zeros((1, 3, 5)))
 ```
 
+```{.python .input}
+#@tab pytorch
+#@save
```
Comment (astonzhang) [inline chapter_computer-vision/anchor.md:314]: why do we save multibox_target to d2l?

File: chapter_computer-vision/anchor.md [hunk 12]
```diff
@@ -210,9 +413,77 @@ output = npx.multibox_detection(
 output
 ```
 
+```{.python .input}
+#@tab pytorch
+#@save
```
Comment (astonzhang) [inline chapter_computer-vision/anchor.md:418]: why do we save non_max_suppression to d2l?

File: chapter_computer-vision/anchor.md [hunk 13]
```diff
@@ -210,9 +413,77 @@ output = npx.multibox_detection(
...
+def non_max_suppression(bounding_boxes: list, iou_threshold: float = 0.1) -> list:
```
Comment (astonzhang) [inline chapter_computer-vision/anchor.md:419]: let's remove type hints, just for consistency

File: chapter_computer-vision/anchor.md [hunk 14]
```diff
@@ -210,9 +413,77 @@ output = npx.multibox_detection(
...
+            if iou > iou_threshold:
+                remove_items.append(bb)
+        bounding_boxes = [bb for bb in bounding_boxes if bb not in remove_items]
+    return filtered_bb
+```
+
+```{.python .input}
+#@tab pytorch
+#@save
+def multibox_detection(id_cat, cls_probs, anchors, nms_threshold):
```
Comment (astonzhang) [inline chapter_computer-vision/anchor.md:440]: why do we save multibox_detection to d2l?

File: chapter_computer-vision/anchor.md [hunk 15]
```diff
@@ -36,16 +45,54 @@ Y = npx.multibox_prior(X, sizes=[0.75, 0.5, 0.25], ratios=[1, 2, 0.5])
 Y.shape
 ```
 
+```{.python .input}
+#@tab pytorch
+#@save
+def multibox_prior(feature_map_sizes, sizes, aspect_ratios):
+    """Compute default box sizes with scale and aspect transform."""
+    sizes = [s*728 for s in sizes]
```
Comment (astonzhang) [inline chapter_computer-vision/anchor.md:53]: put space around *

File: chapter_computer-vision/anchor.md [hunk 16]
```diff
@@ -191,17 +377,34 @@ cls_probs = np.array([[0] * 4,  # Predicted probability for background
...
 
-```{.python .input  n=12}
+```{.python .input}
+fig = d2l.plt.imshow(img)
+show_bboxes(fig.axes, anchors * bbox_scale,
+            ['dog=0.9', 'dog=0.8', 'dog=0.7', 'cat=0.9'])
+```
+
+```{.python .input}
+#@tab pytorch
```
Comment (astonzhang) [inline chapter_computer-vision/anchor.md:399]: we shall merge these two code cells

**Comment by astonzhang** [comment]
Thanks. I just reviewed the style. @AnirudhDagar, please be careful when you review the logic of the code.

General comments that apply to the entire PR:

1. Let's remove all the type hints in functions, just for consistency
2. Avoid empty lines unless they absolutely improve readability
3. Please make sure all the functions marked with @save are absolutely necessary.

**Comment by AnirudhDagar** [comment]
Sure! Thanks for the review @astonzhang
I'll look at the whole PR once again after all these changes are made and see what functions need to be saved.
 
As of now functions like `multibox_prior`, these are not built-in to torch/torchvision. Hence we require them to be saved as they are used in 13.5 and 13.7.

@suvaansh please make the changes as suggested above.

**Comment by StevenJokes** [comment]
Why I can't find the newest def in torch.py
https://github.com/d2l-ai/d2l-en/issues/1374

File: chapter_computer-vision/anchor.md [hunk 17]
```diff
@@ -234,11 +514,10 @@ In practice, we can remove prediction bounding boxes with lower confidence level
 ## Exercises
 
 1. Change the `sizes` and `ratios` values in the `multibox_prior` function and observe the changes to the generated anchor boxes.
-1. Construct two bounding boxes with an IoU of 0.5, and observe their coincidence.
+1. Construct two bounding boxes with and IoU of 0.5, and observe their coincidence.
```
[the remark this answers is not in the document]
Comment (StevenJokes) [reply chapter_computer-vision/anchor.md:517]: agree

File: chapter_computer-vision/anchor.md (hunk 16 above)
Comment (StevenJokes) [reply chapter_computer-vision/anchor.md:399]: We use #tab all to merge @suvaansh

**Comment by AnirudhDagar** [comment]
@suvaansh were you able to make any progress on the review?
Thanks!

**Comment by StevenJokes** [comment]
@AnirudhDagar @astonzhang 
you need to bump `d2l` package version to let me have these newest  `def` in torch.py

**Comment by AnirudhDagar** [comment]
@suvaansh are you still working on this?

**Comment by StevenJokes** [comment]
@suvaansh are you still working on this?

**Comment by AnirudhDagar** [comment]
Closing this for #1565
